in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+  simulate(userRequest: UserRequest): Promise<SimulationOutput>;
+  read(userRequest: UserRequest): Promise<SimulationOutput>;
+  prove(userRequest: UserRequest): Promise<UserRequest>;
+  send(userRequest: UserRequest): Promise<SentTx>;
+}
+
+export type Wallet = UserAPI & AccountInterface & PXE & AccountKeyRotationInterface;
+
+// Swap `EntrypointInterface` for `TxExecutionRequestComponent`
+export interface AccountInterface extends AuthWitnessProvider, TxExecutionRequestComponent {
+  getCompleteAddress(): CompleteAddress;
+  getAddress(): AztecAddress;
+  getChainId(): Fr;
+  getVersion(): Fr;
+
+}
+
+// unchanged
+export interface AuthWitnessProvider {
+  createAuthWit(
+    messageHashOrIntent:
+      | Fr
+      | Buffer
+      | {
+          caller: AztecAddress;
+          action: ContractFunctionInteraction | FunctionCall;
+          chainId?: Fr;
+          version?: Fr;
+        },
+  ): Promise<AuthWitness>;
+}
+
+```
+
+
+## Implementation
+
+### Expose an explicit `proveTx` on PXE
+
+We currently don't have a way to *only* prove a transaction; everything must be simulated first.
+
+So we will expose a `proveTx` method on the PXE that will take a `TxExecutionRequest` and an `ExecutionResult` and return a `Tx`.
+
+### Drop `isFeePayer` from `FeeEntrypointPayload`
+
+
+Removing this makes the code simpler and more flexible: account contracts can just check if fee payer has been set and set it if not.
+
+Concerned users can just inspect the simulation result to see if they are paying the fee.
+
+Then we just have `EntrypointPayload` as a single, concrete class.
+
+### BaseWallet.getTxExecutionRequest
+
+Consider that we have, e.g.:
+
+```ts
+{
+  calls: [{
+    contractInstance: bananaCoinInstance,
+    functionName: 'transfer',
+    args: { 
+      from: aliceAddress,
+      to: bobAddress,
+      value: privateBalance,
+      nonce: 0n
+    },
+  }],
+  paymentMethod
+}
+```
+
+We need to ultimately get to a `TxExecutionRequest`.
+
+#### Translate a `contractInstance`, `functionName` and `args` to a `FunctionCall`
+
+Define helpers somewhere as:
+
+```ts
+
+function findFunctionAbi(contractArtifact: ContractArtifact, functionName: string): FunctionAbi {
+  const functionAbi = contractArtifact.abi.find((abi) => abi.name === functionName);
+  if (!functionAbi) {
+    throw new Error(`Function ${functionName} not found in contract artifact`);
+  }
+  return functionAbi;
+}
+
+function makeFunctionCall(
+  functionAbi: FunctionAbi,
+  instanceAddress: AztecAddress,
+  args: any,
+): FunctionCall {
+  return FunctionCall.from({
+    name: functionAbi.name,
+    args: mapArgsObjectToArray(functionAbi.parameters, args),
+    selector: FunctionSelector.fromNameAndParameters(functionAbi.name, functionAbi.parameters),
+    type: functionAbi.functionType,
+    to: instanceAddress,
+    isStatic: functionAbi.isStatic,
+    returnTypes: functionAbi.returnTypes,
+  });
+}
+
+```
+
+#### main function calls
```

Comment:
The "main" part is a little weird to me. If you are developing, I would just see it as adding a function call not a main one, what is the mental model difference? To distinguish between inner calls being made? But a user would not be doing those anyway 🤷

Reply:
Yeah it's to distinguish between things that would be added by, say, deployment/registration. But if we ditch those options as suggested above then "main" is just the function you're calling from the entrypoint (app payload).